consensus/src/agreement/step.rs

```diff
@@ -130,54 +107,37 @@ impl<D: Database> Executor<D> {
             future_msgs.lock().await.drain_events(self.ru.round, 0)
         {
             for msg in messages {
-                self.collect_inbound_msg(&mut acc, msg).await;
+                self.collect_inbound_msg(msg).await;
             }
         }
 
-        // event_loop for agreements messages
+        // msg_loop for agreement messages
         loop {
-            select! {
-                biased;
-                 // Process the output message from the Accumulator
-                 result = collected_votes_rx.recv() => {
-                    if let Some(aggrements) = result {
-                        if let Some(block) = self.collect_votes(aggrements).await {
-                            // Winning block of this round found.
-                            future_msgs.lock().await.clear_round(self.ru.round);
-                            break Ok(block)
-                        }
+            // Process messages from outside world
+            if let Ok(msg) = self.inbound_queue.recv().await {
+                match msg.header.compare_round(self.ru.round) {
+                    Status::Future => {
```

Comment:
Future Agreement messages are akin to receiving a future block. We should probably trigger a request_block message to the sender so that a sync procedure is initiated if appropriate.
That is, we don't need to run the whole round if we already received the relative Agreement message.

This is probably for a separate PR

Reply:
Not sure what should be the efficient approach here. Maybe we can file another issue to reconsider this.